### PR TITLE
Feature/remove convolver

### PR DESCRIPTION
**Pull request by Jammy2211**
The `Convolver` object mapped out the 2D convolution calculation for a 2D mask and 2D kernel, using the fact that for model-fitting both quantities were fixed and therefore the exact sequence of calculations required could be precomputed in memory.

This object relies heavily on in-place memory manipulation and therefore is not suitable for JAX, therefore I have removed it.

All 2D convolutions are now performed using standard `jax.scipy` convolution methods on 2D arrays.

The `Conolver` object worked on the contents of masked arrays mapped to 1D representation, and the current JAX implementation of convolutions requires mapping back and forth from 1D and 2D. This likely leads to a loss of performance and we may need to optimize these functions in the future.

### CODE REVIEW

File: autoarray/dataset/imaging/dataset.py [hunk 1]
```diff
@@ -167,6 +166,9 @@ def __init__(
 
         self.psf = psf
 
+        if psf.mask.shape[0] % 2 == 0 or psf.mask.shape[1] % 2 == 0:
+            raise exc.KernelException("Kernel2D Kernel2D must be odd")
```
Comment (CKrawczyk) [inline autoarray/dataset/imaging/dataset.py:170]: I have seen this before but never really understood, why do you want the PSF kernel to be odd?

Comment (Jammy2211) [reply autoarray/dataset/imaging/dataset.py:170]: For real space convolution, even x even kernels can induce a half pixel shift depending on where you centre it. This meant that things like the lens model centres had a half pixel shift, in the end, I think.

Alternatively, you could sample the "peak" of the even x even PSF in the central 4 pixels, which means you don't fully sample its shape where it matters most.

This may not matter for an FFT? And certainly wont matter once we do PSF over sampling?

Its definitely a hack I am happy to see removed.

File: autoarray/structures/arrays/kernel_2d.py [hunk 2]
```diff
@@ -481,33 +485,90 @@ def convolved_array_from(self, array: Array2D) -> Array2D:
...
-        ------
-        KernelException if either Kernel2D psf dimension is odd
+        expanded_array_native = jnp.zeros(image.mask.shape)
+
+        expanded_array_native = expanded_array_native.at[slim_to_native].set(
+            image.array
+        )
+        expanded_array_native = expanded_array_native.at[slim_to_native_blurring].set(
+            blurring_image.array
+        )
```
Comment (CKrawczyk) [inline autoarray/structures/arrays/kernel_2d.py:519]: Might be worth checking if `jnp.where` can ends up doing this faster after a `jit`?  Or if it is the same speed, you can reduce this to two lines of code (just double check the lines below to make sure they do the same thing).

```python
expanded_array_native = jnp.where(image.mask.array, 0, image.array)
expanded_array_native = jnp.where(blurring_image.mask.array, expanded_array_native, blurring_image.array)
```

Comment (CKrawczyk) [reply autoarray/structures/arrays/kernel_2d.py:519]: I just double checked this, and I think the index ordering when using `jnp.where` is not quite the same as what you have, so my suggested change might not work as I was expecting.

File: autoarray/structures/arrays/kernel_2d.py [hunk 3]
```diff
@@ -481,33 +485,90 @@ def convolved_array_from(self, array: Array2D) -> Array2D:
...
 
-        convolved_array_2d = scipy.signal.convolve2d(array, self.native, mode="same")
+        slim_to_native = jnp.nonzero(jnp.logical_not(mask.array), size=image.shape[0])
 
-        convolved_array_1d = array_2d_util.array_2d_slim_from(
-            mask_2d=np.array(mask),
-            array_2d_native=np.array(convolved_array_2d),
+        expanded_array_native = jnp.zeros(mask.shape)
+
+        expanded_array_native = expanded_array_native.at[slim_to_native].set(image)
```
Comment (CKrawczyk) [inline autoarray/structures/arrays/kernel_2d.py:552]: As above this can be done with (again, this should be check to make sure it is doing the same thing).

```python
expanded_array_native = jnp.where(mask.array.T, 0, image)
```

File: autoarray/structures/arrays/kernel_2d.py (hunk 2 above)
Comment (CKrawczyk) [reply autoarray/structures/arrays/kernel_2d.py:519]: And with more testing, I have found since we have a 2D mask and a 1D slim from the `jnp.where` does not work as intended.  Keep the code as you have it.

Comment (Jammy2211) [reply autoarray/structures/arrays/kernel_2d.py:519]: Cool, we may be able to produce images in 2D, perform convolution and then flatten.

Storing everything in 1D throughout the whole likelhood function was optimal for `numba` because we could store in memory a lot of arrays that did the mappings for us. But I don't think this is very JAX-y.
